core/src/main/java/io/parsingdata/metal/data/ParseState.java

```diff
@@ -56,12 +55,12 @@ public static ParseState createFromByteStream(final ByteStream input) {
     }
 
     public ParseState addBranch(final Token token) {
-        return new ParseState(order.addBranch(token), source, offset, token.isIterable() ? iterations.add(new Pair<>(token, ZERO)) : iterations);
+        return new ParseState(order.addBranch(token), source, offset, token.isIterable() ? iterations.add(new ImmutablePair<>(token, ZERO)) : iterations);
     }
 
     public ParseState closeBranch(final Token token) {
-        if (token.isIterable() && !iterations.head.getKey().equals(token)) {
-            throw new IllegalStateException(String.format("Cannot close branch for iterable token %s. Current iteration state is for token %s.", token.name, iterations.head.getKey().name));
+        if (token.isIterable() && !iterations.head.left.equals(token)) {
+            throw new IllegalStateException(String.format("Cannot close branch for iterable token %s. Current iteration state is for token %s.", token.name, iterations.head.left.name));
```

Comment:
I'm wondering if `token.name` is enough here. It is not unusual for `Token`s to have no name at all. This `Exception` would become pretty unclear in that case. We could add the `getClass().getSimpleName()` to the message.